File: Common_Concepts/permutation.rb
require 'byebug'

class Array
  def permutations
    return [self] if self.length == 1

    perms = self[1..-1].permutations
    first_num = [self.first]
    total_perms = []

    perms.each do |perm|
      (0..perm.length).each do |id|
        first_half = perm[0...id]
        second_half = perm[id..-1]
        total_perms << first_half + first_num + second_half
      end
    end
    total_perms
  end
end

# class String
#   def permutations
#     # debugger
#     return [self] if length == 1
#     letters = self.chars
#
#     perms = letters[0..-2].permutations
#     last_char = [letters.last]
#     total_perms = []
#
#     perms.each do |perm|
#       (0..perm.length).each do |id|
#         first_half = perm[0...id]
#         second_half = perm[id..-1]
#         total_perms << first_half + last_char + second_half
#       end
#     end
#
#     total_perms
#   end
# end

# def split_string(string)
#   debugger
#   separated = string.chars
#   total_perms = []
#   def permutations(arr, total_perms)
#     return [arr] if arr.length == 1
#
#     perms = permutations(arr[1..-1], total_perms)
#     first_num = [arr[0]]
#
#     perms.each do |perm|
#       (0..perm.length).each do |id|
#         first_half = perm[0...id]
#         second_half = perm[id..-1]
#         total_perms << first_half + first_num + second_half
#       end
#     end
#     total_perms
#   end
#   permutations(separated, total_perms)
#   total_perms
# end

p [1, 2, 3, 4].permutations
# p "help".permutations
# p split_string('hello')

File: Common_Concepts/permutation.js
Array.prototype.permutations = function () {
  if (this.length == 1) return [this];

  let perms = this.slice(0, this.length - 1).permutations();
  let lastNum = this[this.length - 1];
  let totalPerms = [];

  perms.forEach( perm => {
    for (let i = 0; i <= perm.length; i++) {
      let firstHalf = perm.slice(0, i);
      let secondHalf = perm.slice(i);
      totalPerms.push(firstHalf.concat([lastNum], secondHalf));
    }
  })
  return totalPerms;
};

// Array.prototype.permutations = function () {
//   if (this.length == 1) return [this];
//
//   let perms = this.slice(1).permutations();
//   let first = [this[0]];
//   let totalPerms = [];
//   for (let perm of perms) {
//     for (let i = 0; i <= perm.length; i++) {
//       let firstHalf = perm.slice(0, i);
//       let secondHalf = perm.slice(i);
//       totalPerms.push(firstHalf.concat(first, secondHalf));
//     }
//   }
//
//   return totalPerms;
// };

console.log([1,2,3].permutations());
